File: backend/app/schemas/review.py
from datetime import datetime
from enum import Enum
from typing import Any

from pydantic import BaseModel, Field


class Severity(str, Enum):
    critical = "Critical"
    high = "High"
    medium = "Medium"
    low = "Low"


class ReviewIssue(BaseModel):
    line: int | None = None
    type: str
    severity: Severity
    message: str
    suggested_fix: str
    source: str = "local"


class ReviewSummary(BaseModel):
    critical: int = 0
    high: int = 0
    medium: int = 0
    low: int = 0
    score: int = 100


class RefactorSuggestion(BaseModel):
    before: str
    after: str
    reason: str


class ReviewRequest(BaseModel):
    code: str = Field(min_length=1)
    filename: str | None = None
    language: str = "python"
    include_project_context: bool = False
    context_text: str | None = None
    dependency_manifest: str | None = None
    manifest_type: str | None = None


class ReviewResponse(BaseModel):
    issues: list[ReviewIssue]
    summary: ReviewSummary
    technical_debt: str
    overall_assessment: str
    refactor_suggestions: list[RefactorSuggestion] = Field(default_factory=list)


class ReviewRunResponse(ReviewResponse):
    submission_id: int
    created_at: datetime
    provider: str


class SubmissionResult(BaseModel):
    id: int
    filename: str | None
    language: str
    quality_score: int
    created_at: datetime
    review_result: dict[str, Any]

    model_config = {"from_attributes": True}


class ScorePoint(BaseModel):
    created_at: datetime
    score: int


class DashboardMetrics(BaseModel):
    issue_distribution: dict[str, int]
    score_trend: list[ScorePoint]
    submissions: int
    average_score: float

File: frontend/src/types/review.ts
export type Severity = "Critical" | "High" | "Medium" | "Low";

export interface ReviewIssue {
  line: number | null;
  type: string;
  severity: Severity;
  message: string;
  suggested_fix: string;
  source: string;
  confidence?: string;
  original_code?: string | null;
  fixed_code?: string | null;
}

export interface ReviewSummary {
  critical: number;
  high: number;
  medium: number;
  low: number;
  score: number;
}

export interface RefactorSuggestion {
  before: string;
  after: string;
  reason: string;
}

export interface PerformanceHotspot {
  line: number | null;
  operation: string;
  estimated_complexity: string;
  recommendation: string;
  source: string;
}

export interface PerformanceProfile {
  time_complexity: string;
  space_complexity: string;
  confidence: "low" | "medium" | "high" | string;
  hotspots: PerformanceHotspot[];
}

export interface ReviewResponse {
  issues: ReviewIssue[];
  summary: ReviewSummary;
  technical_debt: string;
  overall_assessment: string;
  refactor_suggestions: RefactorSuggestion[];
  performance?: PerformanceProfile;
  provider: string;
  submission_id: number;
  created_at: string;
}

export interface ReviewRequest {
  code: string;
  filename?: string;
  language: string;
  include_project_context?: boolean;
  context_text?: string;
  dependency_manifest?: string;
  manifest_type?: string;
}

export interface ScorePoint {
  created_at: string;
  score: number;
}

export interface DashboardMetrics {
  issue_distribution: Record<Severity, number>;
  score_trend: ScorePoint[];
  submissions: number;
  average_score: number;
}

export type ReviewActionType = "accept_fix" | "ignore_fix" | "accept_issue" | "ignore_issue";

export interface ReviewAction {
  id: number;
  submission_id: number;
  action_type: ReviewActionType;
  item_key: string;
  payload: Record<string, unknown>;
  created_at: string;
}

export interface QuizQuestion {
  question: string;
  options: string[];
  correct_option: number;
  explanation: string;
}

export interface LearningQuizResponse {
  concept: string;
  source: string;
  questions: QuizQuestion[];
}

export interface LearningQuizRequest {
  concept?: string;
  submission_id?: number;
}

export interface LearningQuizSubmitRequest {
  score: number;
  total: number;
}

export interface LearningQuizSubmitResponse {
  score: number;
  total: number;
  quiz_score: number;
}
